### PR TITLE
Add Cubed sphere cell-centers grid

### PR DESCRIPTION
**Pull request by svahl991**
These changes add the ability to create a cubed sphere grid using cell-center points. (Previously the code only supported grids using nodal points.)

It also adds appropriate tests for the new grid type and updates existing tests.

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](http://claassistant.ecmwf.int:80/ecmwf/atlas?pullRequest=69) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](http://claassistant.ecmwf.int:80/ecmwf/atlas?pullRequest=69) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](http://claassistant.ecmwf.int:80/check/ecmwf/atlas?pullRequest=69) it.</sub>

**Comment by wdeconinck** [comment]
Thank you @svahl991 I will have a look as soon as possible. I am currently on leave.

File: src/tests/grid/test_cubedsphere.cc [hunk 1]
```diff
@@ -40,70 +40,80 @@ CASE( "cubedsphere_tile_test" ) {
...
-                std::vector<PointXY> coordinates_2;
-                std::vector<PointXY> coordinates_3;
-                {
-                    for ( auto crd : g.xy() ) {
-                        coordinates_1.push_back( crd );
-                    }
+            SECTION( grid_name ) {
+                if( grid_name == "CS-LFR-L-1" ) {
+                    Log::error() << eckit::Colour::red << "TODO: Fix me!!!. Skipping..." << eckit::Colour::reset << std::endl;
+                    continue;
```
Comment (wdeconinck) [inline src/tests/grid/test_cubedsphere.cc:52]: I added more grids here, and detected that "CS-LFR-L-1" gives segmentation faults, hence it is skipped, but should be fixed soon.
I had already detected this problem before this PR so it is unrelated and should not stand in the way of merging this.

File: src/tests/grid/test_grids.cc [hunk 2]
```diff
@@ -428,107 +428,192 @@ CASE( "test_structured_from_config" ) {
...
+            "CS-ED-C-" + std::to_string( resolution ),
+            "CS-LFR-C-" + std::to_string( resolution )
+        };
+
+        for ( const std::string& grid_name : grid_names ) {
+            Log::info() << "Testing " << grid_name << std::endl;
+            if( grid_name == "CS-LFR-L-1" ) {
+                Log::error() << eckit::Colour::red << "TODO: Fix me!!!. Skipping..." << eckit::Colour::reset << std::endl;
+                continue;
+            }
```
Comment (wdeconinck) [inline src/tests/grid/test_grids.cc:462]: I added more grids and resolutions here, and detected that "CS-LFR-L-1" gives segmentation faults, hence it is skipped, but should be fixed soon. It must be related to previous comment.

**Review by wdeconinck — APPROVED** [review]

**Comment by wdeconinck** [comment]
@svahl991 looks all very good! thanks!
<hr>
@MarekWlasak I highlighted in my review above a preexisting issue about the resolution "CS-LFR-L-1" which probably best needs fixing in a separate PR.
<hr>
@odlomax also following fails still, and probably yet another PR, or solved together with your following meshgen PR.

```
atlas-meshgen CS-LFR-L-2. # fails with any N
atlas-meshgen CS-EA-L-1.  # fails with N=1 only  (fv3 tiling)
atlas-meshgen CS-ED-L-1.  # fails with N=1 only  (fv3 tiling)
```
Output for `atlas-meshgen CS-LFR-L-2`:
```
=========================================
[0] TERMINATING atlas-meshgen
-----------------------------------------
ArrayView index 2147483647 out of bounds: 2147483647 >= 54
-----------------------------------------
LOCATION:  (/Users/willem/work/atlas-bundle/source/atlas/src/atlas/array/ArrayUtil.cc +22 throw_OutOfRange)
-----------------------------------------
BACKTRACE
-----------------------------------------
backtrace [1] stack has 17 addresses
 0   libeckit.dylib                      0x0000000107daa264 eckit::BackTrace::dump()+ 
 1   libeckit.dylib                      0x0000000107c27868 eckit::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, eckit::CodeLocation const&)+ 
 2   libeckit.dylib                      0x0000000107c27ac8 eckit::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, eckit::CodeLocation const&)+ 
 3   libatlas.0.25.dylib                 0x0000000105b03ce0 atlas::throw_Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, eckit::CodeLocation const&)+ 
 4   libatlas.0.25.dylib                 0x000000010534933c atlas::array::throw_OutOfRange(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char, int, int)+ 
 5   libatlas.0.25.dylib                 0x00000001056b6310 void atlas::array::ArrayView<double, 2>::check_bounds_part<0, int, atlas::XYZ>(int, atlas::XYZ) const+ 
 6   libatlas.0.25.dylib                 0x00000001056b6228 void atlas::array::ArrayView<double, 2>::check_bounds<int, atlas::XYZ>(int, atlas::XYZ) const+ 
 7   libatlas.0.25.dylib                 0x00000001056aca38 double& atlas::array::ArrayView<double, 2>::operator()<int, atlas::XYZ>(int, atlas::XYZ)+ 
 8   libatlas.0.25.dylib                 0x000000010574de40 atlas::meshgenerator::CubedSphereMeshGenerator::generate(atlas::Grid const&, atlas::grid::Distribution const&, atlas::Mesh&) const::$_1::operator()(std::__1::array<long, 3ul>, std::__1::array<long, 3ul>) const+ 
 9   libatlas.0.25.dylib                 0x000000010574cc7c atlas::meshgenerator::CubedSphereMeshGenerator::generate(atlas::Grid const&, atlas::grid::Distribution const&, atlas::Mesh&) const+ 
 10  libatlas.0.25.dylib                 0x0000000105776398 atlas::meshgenerator::MeshGeneratorImpl::generate(atlas::Grid const&, atlas::grid::Partitioner const&, atlas::Mesh&) const+ 
 11  libatlas.0.25.dylib                 0x0000000105776424 atlas::meshgenerator::MeshGeneratorImpl::generate(atlas::Grid const&, atlas::grid::Partitioner const&) const+ 
 12  libatlas.0.25.dylib                 0x000000010574b408 atlas::MeshGenerator::generate(atlas::Grid const&, atlas::grid::Partitioner const&) const+ 
 13  atlas-meshgen                       0x0000000104ec17dc Meshgen2Gmsh::execute(eckit::option::CmdArgs const&)+ 
 14  libatlas.0.25.dylib                 0x00000001052fceec atlas::AtlasTool::start()+ 
 15  atlas-meshgen                       0x0000000104ec2ff0 main+ 
 16  libdyld.dylib                       0x0000000183530420 start+ 

end of backtrace dump ...
=========================================
```

**Comment by odlomax** [comment]
> @svahl991 looks all very good! thanks!
> 
> @MarekWlasak I highlighted in my review above a preexisting issue about the resolution "CS-LFR-L-1" which probably best needs fixing in a separate PR.
> @odlomax also following fails still, and probably yet another PR, or solved together with your following meshgen PR.
> ```
> atlas-meshgen CS-LFR-L-2. # fails with any N
> atlas-meshgen CS-EA-L-1.  # fails with N=1 only  (fv3 tiling)
> atlas-meshgen CS-ED-L-1.  # fails with N=1 only  (fv3 tiling)
> ```
> 
> Output for `atlas-meshgen CS-LFR-L-2`:
> 
> ```
> =========================================
> [0] TERMINATING atlas-meshgen
> -----------------------------------------
> ArrayView index 2147483647 out of bounds: 2147483647 >= 54
> -----------------------------------------
> LOCATION:  (/Users/willem/work/atlas-bundle/source/atlas/src/atlas/array/ArrayUtil.cc +22 throw_OutOfRange)
> -----------------------------------------
> BACKTRACE
> -----------------------------------------
> backtrace [1] stack has 17 addresses
>  0   libeckit.dylib                      0x0000000107daa264 eckit::BackTrace::dump()+ 
>  1   libeckit.dylib                      0x0000000107c27868 eckit::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, eckit::CodeLocation const&)+ 
>  2   libeckit.dylib                      0x0000000107c27ac8 eckit::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, eckit::CodeLocation const&)+ 
>  3   libatlas.0.25.dylib                 0x0000000105b03ce0 atlas::throw_Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, eckit::CodeLocation const&)+ 
>  4   libatlas.0.25.dylib                 0x000000010534933c atlas::array::throw_OutOfRange(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char, int, int)+ 
>  5   libatlas.0.25.dylib                 0x00000001056b6310 void atlas::array::ArrayView<double, 2>::check_bounds_part<0, int, atlas::XYZ>(int, atlas::XYZ) const+ 
>  6   libatlas.0.25.dylib                 0x00000001056b6228 void atlas::array::ArrayView<double, 2>::check_bounds<int, atlas::XYZ>(int, atlas::XYZ) const+ 
>  7   libatlas.0.25.dylib                 0x00000001056aca38 double& atlas::array::ArrayView<double, 2>::operator()<int, atlas::XYZ>(int, atlas::XYZ)+ 
>  8   libatlas.0.25.dylib                 0x000000010574de40 atlas::meshgenerator::CubedSphereMeshGenerator::generate(atlas::Grid const&, atlas::grid::Distribution const&, atlas::Mesh&) const::$_1::operator()(std::__1::array<long, 3ul>, std::__1::array<long, 3ul>) const+ 
>  9   libatlas.0.25.dylib                 0x000000010574cc7c atlas::meshgenerator::CubedSphereMeshGenerator::generate(atlas::Grid const&, atlas::grid::Distribution const&, atlas::Mesh&) const+ 
>  10  libatlas.0.25.dylib                 0x0000000105776398 atlas::meshgenerator::MeshGeneratorImpl::generate(atlas::Grid const&, atlas::grid::Partitioner const&, atlas::Mesh&) const+ 
>  11  libatlas.0.25.dylib                 0x0000000105776424 atlas::meshgenerator::MeshGeneratorImpl::generate(atlas::Grid const&, atlas::grid::Partitioner const&) const+ 
>  12  libatlas.0.25.dylib                 0x000000010574b408 atlas::MeshGenerator::generate(atlas::Grid const&, atlas::grid::Partitioner const&) const+ 
>  13  atlas-meshgen                       0x0000000104ec17dc Meshgen2Gmsh::execute(eckit::option::CmdArgs const&)+ 
>  14  libatlas.0.25.dylib                 0x00000001052fceec atlas::AtlasTool::start()+ 
>  15  atlas-meshgen                       0x0000000104ec2ff0 main+ 
>  16  libdyld.dylib                       0x0000000183530420 start+ 
> 
> end of backtrace dump ...
> =========================================
> ```

The mesh generator in this branch is _very_ limited. It can only work with FV3 tiling and it doesn't surprise me that it fails with N=1. I can add some exception handing to catch these cases if it helps.

The mesh generator I'm currently working on only works with "C" grids with N > 1. It also requires that the `Tiles` class has a `CubedSpherePeriodicity` method. I think `LFRicTiles` is the only one this this implemented. @MarekWlasak will the FV3 Tiles class correctly throw a `not_implemented` exception if I try to use it?

**Comment by wdeconinck** [comment]
Hi @odlomax it would help if we get a specific "known not to work" kind of message rather than an index out of bounds which only happens in Debug builds, and would result in segmentation faults in release builds

**Comment by wdeconinck** [comment]
Merged in develop with commit ec341ac1e281939145c39cbc5c9d63ee23d34df8

**Pull request closed without merging**